src/kubernetes_cluster/proof/controller_runtime_safety.rs

```diff
@@ -46,6 +46,222 @@ pub open spec fn pending_req_has_unique_id<T>(cr_key: ObjectRef) -> StatePred<St
     }
 }
 
+pub open spec fn in_flight_resp_has_lower_resp_id<T>() -> StatePred<State<T>> {
+    |s: State<T>| {
+        forall |resp_msg: Message|
+            #[trigger] s.message_in_flight(resp_msg)
+            && resp_msg.content.is_APIResponse()
+            ==> resp_msg.content.get_resp_id() < s.chan_manager.cur_chan_id
+    }
+}
+
+pub proof fn lemma_always_in_flight_resp_has_lower_resp_id<T>(reconciler: Reconciler<T>)
+    ensures
+        sm_spec(reconciler).entails(always(lift_state(in_flight_resp_has_lower_resp_id()))),
+{
+    let invariant = in_flight_resp_has_lower_resp_id::<T>();
+    let stronger_next = |s, s_prime: State<T>| {
+        next(reconciler)(s, s_prime)
+        && in_flight_req_has_lower_req_id::<T>()(s)
+    };
+    lemma_always_in_flight_req_has_lower_req_id::<T>(reconciler);
+    strengthen_next::<State<T>>(sm_spec(reconciler), next(reconciler), in_flight_req_has_lower_req_id::<T>(), stronger_next);
+    assert forall |s, s_prime: State<T>| invariant(s) && #[trigger] stronger_next(s, s_prime) implies invariant(s_prime) by {
+        assert forall |msg: Message| #[trigger] s_prime.message_in_flight(msg) && msg.content.is_APIResponse() implies msg.content.get_resp_id() < s_prime.chan_manager.cur_chan_id by {
+            assert(s.chan_manager.cur_chan_id <= s_prime.chan_manager.cur_chan_id);
+            if (s.message_in_flight(msg)) {
+                assert(msg.content.get_resp_id() < s.chan_manager.cur_chan_id);
+                assert(msg.content.get_resp_id() < s_prime.chan_manager.cur_chan_id);
+            }
+            else {
+                let next_step = choose |step: Step| next_step(reconciler, s, s_prime, step);
+                assert(next_step.is_KubernetesAPIStep());
+                let input = next_step.get_KubernetesAPIStep_0();
+                assert(input.is_Some());
+                assert(s.message_in_flight(input.get_Some_0()));
+                assert(input.get_Some_0().content.is_APIRequest());
+                assert(msg.content.get_resp_id() == input.get_Some_0().content.get_req_id());
+                assert(input.get_Some_0().content.get_req_id() < s.chan_manager.cur_chan_id);
+            }
+        };
+    };
+    init_invariant::<State<T>>(sm_spec(reconciler), init(reconciler), stronger_next, invariant);
+}
+
+pub open spec fn every_in_flight_req_is_unique<T>() -> StatePred<State<T>> {
```

Comment:
If you write this one as
```
        forall |msg: Message|
            msg.content.is_APIRequest() && s.message_in_flight(msg)
            ==> #[trigger] s.message_in_flight_count(msg) == 1
```
The proof of `lemma_always_every_in_flight_req_is_unique` might be easier.